README.md
# clickhouse-udp-proxy

Simple UDP proxy for ClickHouse.

It accepts UDP packets, send it to Redis Cluster with [Ruthie](https://github.com/kak-tus/ruthie). Then Ruthie write them to ClickHouse.

Proxy role - run near some writer and accept metrics in UDP packets. That can be added at any code (sync or async) without any code refactoring.

Best idea - write to Ruthie directly, but sometimes (in Perl code) it is not possible.

## Configuration

```
PROXY_REDIS_ADDR=172.17.0.1:7000
PROXY_SHARDS_COUNT=10
PROXY_PENDING_BUFFER_SIZE=1000000
PROXY_PIPE_BUFFER_SIZE=50000
```

## Run

```
docker run --rm -it -p 9001:9001/udp -p 9000:9000/tcp kaktuss/clickhouse-udp-proxy
```

## Packet format

```
{"types":["string","string","int"],"data":["2017-09-09","2017-09-09 12:26:03",3],"query":"INSERT INTO test (dt,dt1,id) VALUES (?,?,?);","version":1}
```

version - protocol version for future changes

query - query to run

data - data to query

types - data in JSON like "1" (integers) golang unpacks as float, and this data
fails to insert in integer columns in ClickHouse, so we need to convert this
values.

If you set int type - value will be converted to int. Any other type will be
ignored (they extracted correctly).

Send example see at dev/test.pl

## Healthcheck

Service listens at 9000 TCP port and response to HTTP queries with code 200 or
code 500.


Changes
v2.0.2 2019-11-10
  - Use latest Ami and Ruthie.
  - Use go modules.
  - Remove vendoring.
  - Remove packages from private repo.

v2.0.1 2018-10-28
  - Update Ami version.
  - Remove PROXY_CONSUMER_NAME and PROXY_PREFETCH_COUNT from docker image.

v2.0.0 2018-10-28
  - Swithed to Ruthie - Redis Cluster based writer.

v1.0.3 2018-10-17
  - Fix memory leak.

v1.0.2 2018-10-17
  - Fix possible send deadlock.
  - Bigger possible memory usage.

v1.0.1 2018-10-06
  - Fix send bug.

v1.0.0 2018-10-06
  - Update Alpine to 3.8.
  - Use app environment.
  - Remove direct ClickHouse support, write to CH with Corrie.
  - Use semver.

0.18 2018-03-06
  - More stable healthcheck.

0.17 2018-03-06
  - HTTP healthcheck support.

0.16 2018-03-06
  - Use rollback (fixed in latest version of CH library) to prevent memory leak.

0.15 2018-03-02
  - Disable profiling, memory leak found in CH library.
  - Correct program stop.

0.14 2018-03-02
  - Temporary enable profiling.

0.13 2018-03-02
  - Bigger write timeout.

0.12 2018-03-01
  - Fix memory leak.

0.11 2017-12-29
  - Fix possible memory leak.

0.10 2017-12-20
  - Fix panic of type conversion.
  - Log error on incompatible got and waited type.

0.9 2017-12-01
  - Not log in normal mode to stderr.

0.8 2017-11-25
  - Bigger batch size.

0.7 2017-11-25
  - Default configuration to usual CH CPU load.

0.6 2017-11-25
  - Not send (and start transaction) with no values for query.

0.5 2017-11-25
  - Better error logging.
  - Logging count of inserted values.

0.4 2017-11-24
  - Configurable batch and period.

0.3 2017-09-10
  - Small fixes in error processing.

0.2 2017-09-10
  - Resend in period of 1 sec. if we have few data.
  - More items at once (100->1000).

0.1 2017-09-09
  - First release.


dev/test.pl
#!/usr/bin/env perl

use strict;
use warnings;
use v5.10;
use utf8;

use Cpanel::JSON::XS qw(encode_json);
use IO::Socket;

my $sock = IO::Socket::INET->new(
  Proto    => 'udp',
  PeerPort => 9001,
  PeerAddr => '172.17.0.1',
) or die "Could not create socket: $!\n";

for ( 1 .. 100 ) {
  my %val = (
    query   => 'INSERT INTO test (dt_part,dt,id,fl) VALUES (?,?,?);',
    data    => [ "2017-09-09", "2017-09-09 12:26:03", $_, 1.23 ],
    types   => [ "string", "string", "int", "float" ],
    version => 1,
  );

  $sock->send( encode_json( \%val ) ) or die "Send error: $!\n";

  sleep 1;
}


main.go
package main

import (
	"net/http"
	"os"
	"os/signal"

	"github.com/kak-tus/clickhouse-udp-proxy/listener"
	"github.com/kak-tus/healthcheck"
	"go.uber.org/zap"
)

func main() {
	logger, err := zap.NewProduction()
	if err != nil {
		panic(err)
	}

	log := logger.Sugar()

	cnf, err := newConfig()
	if err != nil {
		log.Panic(err)
	}

	hlth := healthcheck.NewHandler()
	hlth.Add("/healthcheck", func() (healthcheck.State, string) {
		return healthcheck.StatePassing, "ok"
	})

	lstn, err := listener.NewListener(cnf.Listener, log)
	if err != nil {
		log.Panic(err)
	}

	go func() {
		if err := http.ListenAndServe(cnf.Healthcheck.Listen, hlth); err != nil {
			log.Panic(err)
		}
	}()

	go lstn.Start()

	st := make(chan os.Signal, 1)
	signal.Notify(st, os.Interrupt)

	<-st
	log.Info("Stop")

	lstn.Stop()

	_ = log.Sync()
}


config.go
package main

import (
	"github.com/iph0/conf"
	"github.com/iph0/conf/envconf"
	"github.com/iph0/conf/fileconf"
	"github.com/kak-tus/clickhouse-udp-proxy/listener"
)

type config struct {
	Healthcheck healthcheckConfig
	Listener    listener.Config
}

type healthcheckConfig struct {
	Listen string
}

func newConfig() (*config, error) {
	fileLdr := fileconf.NewLoader("etc", "/etc")
	envLdr := envconf.NewLoader()

	configProc := conf.NewProcessor(
		conf.ProcessorConfig{
			Loaders: map[string]conf.Loader{
				"file": fileLdr,
				"env":  envLdr,
			},
		},
	)

	configRaw, err := configProc.Load(
		"file:clickhouse-udp-proxy.yml",
		"env:^PROXY_",
	)

	if err != nil {
		return nil, err
	}

	var cnf config
	if err := conf.Decode(configRaw, &cnf); err != nil {
		return nil, err
	}

	return &cnf, nil
}


listener/listener.go
package listener

import (
	"net"
	"strings"
	"sync"
	"time"

	"github.com/go-redis/redis"
	jsoniter "github.com/json-iterator/go"
	"github.com/kak-tus/ami"
	"github.com/kak-tus/ruthie/message"
	"go.uber.org/zap"
)

func NewListener(cnf Config, log *zap.SugaredLogger) (*Listener, error) {
	addrs := strings.Split(cnf.Redis.Addrs, ",")

	pr, err := ami.NewProducer(
		ami.ProducerOptions{
			Name:              "ruthie",
			ShardsCount:       cnf.ShardsCount,
			PendingBufferSize: cnf.PendingBufferSize,
			PipeBufferSize:    cnf.PipeBufferSize,
			PipePeriod:        time.Microsecond * 10,
		},
		&redis.ClusterOptions{
			Addrs: addrs,
		},
	)
	if err != nil {
		return nil, err
	}

	lstn := &Listener{
		logger: log,
		m:      &sync.Mutex{},
		config: cnf,
		pr:     pr,
	}

	return lstn, nil
}

// Start listener
func (l *Listener) Start() {
	decoder := jsoniter.Config{UseNumber: true}.Froze()

	conn, err := net.ListenPacket("udp", ":"+l.config.Port)
	if err != nil {
		l.logger.Panic(err)
	}

	buf := make([]byte, 65535)
	var parsed reqType

	l.m.Lock()

	l.logger.Info("Started listener")

	for {
		if l.stop {
			break
		}

		err = conn.SetReadDeadline(time.Now().Add(1 * time.Second))
		if err != nil {
			continue
		}

		num, _, err := conn.ReadFrom(buf)
		if err != nil {
			continue
		}

		err = decoder.Unmarshal(buf[0:num], &parsed)
		if err != nil {
			l.logger.Error(err)
			continue
		}

		body, err := message.Message{
			Query: parsed.Query,
			Data:  parsed.Data,
		}.Encode()

		if err != nil {
			l.logger.Error(err)
			continue
		}

		l.pr.Send(body)
	}

	conn.Close()
	l.m.Unlock()
}

func (l *Listener) Stop() {
	l.logger.Info("Stop listener")

	l.stop = true
	l.m.Lock()
	l.pr.Close()

	l.logger.Info("Stopped listener")
}


proxy.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"reflect"
	"strconv"
	"sync"
	"syscall"
	"time"

	"github.com/kshvakov/clickhouse"
)

type reqType struct {
	Query   string        `json:"query"`
	Data    []interface{} `json:"data"`
	Types   []string      `json:"types"`
	Version int           `json:"version"`
}

var db *sql.DB
var logger = log.New(os.Stdout, "", log.LstdFlags)
var errLogger = log.New(os.Stderr, "", log.LstdFlags)

func main() {
	connectDB()

	ch := make(chan reqType, 100000)
	stopChan := make(chan int)
	go aggregate(ch, stopChan)

	go listen(ch)
	go healthcheck()

	<-stopChan
	logger.Println("Exit")
}

func connectDB() {
	addr := os.Getenv("CLICKHOUSE_ADDR")

	var err error
	db, err = sql.Open("clickhouse", "tcp://"+addr+"?write_timeout=60")
	if err != nil {
		log.Panicln(err)
	}

	err = db.Ping()
	if err != nil {
		exception, ok := err.(*clickhouse.Exception)
		if ok {
			log.Panicln(fmt.Sprintf("[%d] %s \n%s", exception.Code, exception.Message, exception.StackTrace))
		} else {
			log.Panicln(err)
		}
	}

	return
}

func listen(ch chan reqType) {
	conn, err := net.ListenPacket("udp", ":9001")
	if err != nil {
		log.Panicln(err)
	}

	buf := make([]byte, 65535)
	mutex := &sync.Mutex{}

	stopSignal := make(chan os.Signal, 1)
	signal.Notify(stopSignal, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-stopSignal
		logger.Println("Got stop signal, stop listening UDP")

		mutex.Lock()
		conn.Close()
		conn = nil
		mutex.Unlock()
	}()

	for {
		mutex.Lock()

		if conn == nil {
			break
		}

		err = conn.SetReadDeadline(time.Now().Add(1 * time.Second))
		if err != nil {
			mutex.Unlock()
			continue
		}

		num, _, err := conn.ReadFrom(buf)
		if err != nil {
			mutex.Unlock()
			continue
		}

		mutex.Unlock()

		var parsed reqType
		err = json.Unmarshal(buf[0:num], &parsed)
		if err != nil {
			errLogger.Println(err)
			continue
		}

		ch <- parsed
	}

	logger.Println("Close channel")
	close(ch)
}

func aggregate(ch chan reqType, stopChan chan int) {
	period, err := strconv.ParseUint(os.Getenv("PROXY_PERIOD"), 10, 64)
	if err != nil {
		errLogger.Println(err)
		period = 5
	}

	batch, err := strconv.ParseUint(os.Getenv("PROXY_BATCH"), 10, 64)
	if err != nil {
		errLogger.Println(err)
		batch = 10000
	}

	parsedVals := make(map[string][]reqType)
	parsedCnts := make(map[string]int)

	start := time.Now()

	for {
		parsed, more := <-ch

		if more {
			if parsedVals[parsed.Query] == nil {
				parsedVals[parsed.Query] = make([]reqType, batch)
			}

			parsedCnts[parsed.Query]++
			parsedVals[parsed.Query][parsedCnts[parsed.Query]-1] = parsed

			if parsedCnts[parsed.Query] >= int(batch) {
				_ = send(parsed.Query, parsedVals[parsed.Query][0:parsedCnts[parsed.Query]])
				logger.Println(fmt.Sprintf("Sended %d values for %q", parsedCnts[parsed.Query], parsed.Query))
				parsedCnts[parsed.Query] = 0
			}
		}

		if time.Now().Sub(start).Seconds() >= float64(period) || !more {
			for k, v := range parsedVals {
				if parsedCnts[k] > 0 {
					_ = send(k, v[0:parsedCnts[k]])
					logger.Println(fmt.Sprintf("Sended %d values for %q", parsedCnts[k], k))
					parsedCnts[k] = 0
				}
			}

			start = time.Now()
		}

		if !more {
			logger.Println("No more messages")
			stopChan <- 1
			break
		}
	}
}

func send(query string, vals []reqType) error {
	tx, err := db.Begin()
	if err != nil {
		errLogger.Println(err)
		return err
	}

	stmt, err := tx.Prepare(query)
	if err != nil {
		tx.Rollback()
		errLogger.Println(err)
		return err
	}

	for _, val := range vals {
		var args []interface{}
		argErr := 0

		for i := 0; i < len(val.Data); i++ {
			if val.Types[i] == "int" && reflect.TypeOf(val.Data[i]).Kind() == reflect.String {
				errLogger.Println(fmt.Sprintf("Got type string, but waited int: %q", val))
				argErr = 1
			} else if val.Types[i] == "int" {
				args = append(args, int64(val.Data[i].(float64)))
			} else {
				args = append(args, val.Data[i])
			}
		}

		if argErr == 0 {
			_, err := stmt.Exec(args...)

			if err != nil {
				tx.Rollback()
				errLogger.Println(err)
				return err
			}
		}
	}

	err = tx.Commit()
	if err != nil {
		errLogger.Println(err)
		return err
	}

	return nil
}

func healthcheck() {
	srv := &http.Server{Addr: ":9001"}

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		for i := 0; i < 10; i++ {
			err := db.Ping()
			if err == nil {
				fmt.Fprintf(w, "ok")
				return
			}

			errLogger.Println(err)
			time.Sleep(time.Second)
		}

		w.WriteHeader(http.StatusInternalServerError)
		fmt.Fprintf(w, "err")
	})

	stopSignal := make(chan os.Signal, 1)
	signal.Notify(stopSignal, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-stopSignal
		logger.Println("Got stop signal, stop listening HTTP")

		err := srv.Shutdown(nil)
		if err != nil {
			errLogger.Println(err)
		}
	}()

	err := srv.ListenAndServe()
	if err != nil && err != http.ErrServerClosed {
		log.Panicln(err)
	}
}
